### ISSUE TITLE
New version online

### ISSUE THREAD

**Issue by niccokunzmann** [niccokunzmann/hanging_threads#15]
There is a new version online: 1.0.0
What should be in the next version?

**Comment by coffenbacher** [comment]
It would be cool to be able to specify certain threads as OK to hang, I have some threads that intentionally block for a while

**Comment by artemrizhov** [comment]
Ignoring certain threads is what I've already started to work on! :) But
I'm still not sure what is the best way to do this. I'm going to do more
experiments.

чт, 19 янв 2017 г., 7:11 coffenbacher <notifications@github.com>:

**Comment by niccokunzmann** [comment]
I ask myself the question: How would I like to use it?

This comes into my mind:

    t = Thread(...)
    hanging_threads.ignore(t)
    hanging_threads.ignore(t.name)

or

    def thread():
        hanging_threads.ignore()

What are your thoughts?

**Comment by coffenbacher** [comment]
Looks good. Actually I'm not sure how to do exactly what I want, because
some of the threads I want to ignore are spawned by packages I'm using... I
haven't looked into how to get pointers to them.

**Comment by artemrizhov** [comment]
The problem is that such threads may be started automatically by
third-party modules. For example Django starts two threads for development
server. I'm trying to ignore the first started threads using timeout, but
this solution looks a bit hacky, and I'm  not sure I really want to ignore
those threads forever because they also potentially may hang.

**Comment by niccokunzmann** [comment]
Without looking into how it should be implemented: How would you like to use it?

    import module
    hanging_threads.stop_monitoring(module)

How to use it is the first step: Specification first. If you can write an empty function with a good documentation, we can find a way to (test and) implement.

Threads are started by functions. They have code objects. There is a lot of information available. The question is: What behavior/interface would you like assuming all is possible. - I like to divide what and how.

**Comment by artemrizhov** [comment]
Maybe we should ask ourself what problem we want to solve? Why we want to
ignore some threads?

>

**Comment by coffenbacher** [comment]
Great point. My problem is that I start ~10 threads, each of which may
hang. However, those threads use libraries that spawn other threads
that *intentionally
*hang on time.sleep. So, perhaps a better solution for my scenario is not
ignoring threads but rather a whitelist of hang locations.

**Comment by artemrizhov** [comment]
> However, those threads use libraries that spawn other threads that *intentionally* hang on time.sleep

Ok, so let them hang. This is not real problem because this is expected behaviour. What is the real problem with these sleeping threads? I mean, we should think why we are bothering about this at all.

**Comment by coffenbacher** [comment]
Right, the sleeping threads are fine. It is the other threads that hang
unintentionally that are the problem, and unfortunately the error messages
from the sleeping threads swamp any information from the other ones.

I think a whitelist of files / lines to ignore would be sufficient for my
scenario. I'm not sure about other scenarios.

On Jan 19, 2017 9:02 AM, "Artem Rizhov" <notifications@github.com> wrote:

However, those threads use libraries that spawn other threads that
*intentionally* hang on time.sleep
Ok, so let them hang. This is not real problem because this is expected
behaviour. What is the real problem with these sleeping threads? I mean, we
should think why we are bothering about this at all.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/niccokunzmann/hanging_threads/issues/15#issuecomment-273834311>,
or mute the thread

**Comment by niccokunzmann** [comment]
How would you like to specify such a hang location?

    start_monitoring(condition=lambda stack:
stack.starts_in_module(__name__))

to only monitor threads from in the module or

    start_monitoring(condition=lambda stack: not
stack.has_call_to(time.sleep))

to not monitor time.sleep() ?

Which interface expresses most clearly what you would like to achieve?
What are your ideas?

**Comment by coffenbacher** [comment]
Good idea, maybe allow the user to pass a function / lambda that returns a
boolean based on whether the hang should be ignored.

Like this pseudocode (sorry on mobile, also don't exactly know what the
argument to the function would contain but something like this

def ignore(stack):
     if 'time.sleep' in str(stack.current_line):
           return True


hanging_threads.start_monitoring(whitelist_func=ignore)

**Comment by artemrizhov** [comment]
> the error messages from the sleeping threads swamp any information from the other ones.

So, this is the real problem, I guess. Maybe we can just get rid of this swamping messages?

My propositions:
1. Rewrite the logging algorithm so that it reports once when a thread hangs, and then reports once again when the thread awakes. This is done already in my branch, I just have to cleanup things and create a PR.
2. Add an option to ignore repeating hangs, so if a thread hangs on certain line and then awakes, then hanging on this line will not be considered as hangs anymore.

**Comment by coffenbacher** [comment]
Sure, those options would work for me as well

**Comment by artemrizhov** [comment]
Ok, I'll prepare PR for the first part (which seems reasonable in any case), and then we will be able to implement the second one, if everybody agree. @niccokunzmann what do you think about this?

**Comment by artemrizhov** [comment]
@coffenbacher It also may be useful to add an option to ignore threads that hangs on `time.sleep` because this is a common case for the threads that we want to ignore. Not sure if this is a good idea to suppress any messages as soon as they are not annoying, so maybe this option is not really useful. Any suggestions?

**Comment by niccokunzmann** [comment]
Sure, please go ahead.

**Comment by niccokunzmann** [comment]
@coffenbacher Would you like to create a pull-request with the documentation for the functionality which you would like to use?

**Comment by artemrizhov** [comment]
I've added #21 and #31, but still not sure if it makes sense to ignore `time.sleep`. One may get hanging threads because of `sleep` forgotten after debugging, or because of too large sleep time. So this function also may be a cause of hanging threads.

**Comment by niccokunzmann** [comment]
If there are a lot of prints, it could be good to have an option to ignore time.sleep in the future.
